sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DefaultSynchronousMetricStorage.java

```diff
@@ -83,8 +85,13 @@ Queue<AggregatorHandle<T, U>> getAggregatorHandlePool() {
 
   @Override
   public void recordLong(long value, Attributes attributes, Context context) {
-    AggregatorHandle<T, U> handle = getAggregatorHandle(attributes, context);
-    handle.recordLong(value, attributes, context);
+    long stamp = sl.readLock();
```

Comment:
I'm switching to `ReentrantReadWriteLock` for reasons described [here](https://github.com/open-telemetry/opentelemetry-java/pull/5932#issuecomment-1775734632).

> We must never block upon recording, as we deteriorate performance (recording must wait for write lock to finish).

Sure. The approach you outline is optimistic reads. There's a number of different ways to do this, and they appear to be simpler / higher performance with StampedLock. But in either case, I'll rework it so recording measurements doesn't have to block.

Reply:
We're still blocking :) Once a collection start you grab a writeLock, which blocks all readLock until collection has finished. Hence I suggested a different approach outlined above.

Reply:
The logic you suggest does check out and does not block. I'm running the performance tests now to evaluate if reading a volatile variable every time we record on the hot path degrades performance in a serious way. It could be the case that reading a non-volatile variable and only blocking for an extremely short amount of time once per collection is better than reading a volatile variable every record but never blocking.

Reply:
Ok so here's the JMH results for the [plain read write lock solution which blocks for a narrow window when collecting](https://gist.github.com/jack-berg/0c74bac7eb881e3a181346dfacdb603a) vs. @asafm's proposal [to always read from a volatile but never block](https://gist.github.com/jack-berg/b614ef5dc9f6bacafc8097afd8f5ca87). Source code for the volatile, never blocking solution [here](https://github.com/open-telemetry/opentelemetry-java/compare/main...jack-berg:opentelemetry-java:delta-concurrency-bug-volatile?expand=1).

The solution to always read from a volatile and never block reduces performance on the record path by a modest ~4% versus the read write lock approach which blocks briefly during collection. Its also worth noting that as implemented, the volatile never block approach impacts the cumulative performance as well, which isn't strictly necessary since this concurrency bug only affects delta. Cumulative should be able to read a non-volatile variable safely since it never needs to change it.